### DISCUSSION TITLE
new protocol support：KCP

### DISCUSSION THREAD

**Discussion by symsimmy** [apache/apisix#5811] [Ideas]
We are developing PVP games using KCP （a reliable UDP protocol）, which is widely used in real-time battle games and we hope apisix will add support for that.  Or if it's not convenient so far，could  you please let me know how to develope custom plugins to support this feature? What is the general idea of development? Thank you very much！

In addition, our game server is developed in C#, currently there is no C# runner, We would appreciate it if you could include it in your development plan.

kcp repo url：https://github.com/skywind3000/kcp

**Comment by moonming** [comment]
@membphis please take a look

**Comment by tokers** [comment]
@symsimmy I think you may refer to the mqtt plugin implementation, see https://github.com/apache/apisix/blob/master/apisix/stream/plugins/mqtt-proxy.lua.

This is a good example about how we can support other protocols with the APISIX layer-4 proxy.

**Comment by membphis** [comment]
@symsimmy I read the doc of https://github.com/skywind3000/kcp .

I have some questions:

1. Do you want to implement kcp proxy? like this: client(kcp) -> (UDP) -> APISIX(kcp proxy) -> (UDP) -> server(UDP)
2. Or just needs kcp server? client(kcp) -> (UDP) -> APISIX(kcp server, response directly)

waiting for your response.

**Reply by symsimmy** [reply]
i want to use apisix as game gateway,so the route would be 1st one:
 like this: : client(kcp) -> (UDP) -> APISIX(kcp proxy) -> (UDP) -> server(kcp)

**Reply by membphis** [reply]
ok, got it.

Do you need a 4-layer UDP proxy or a 7-layer KCP protocol proxy?

If you need a 7-layer KCP proxy, the APISIX need to parse the KCP request.

if we chat together in the weekly meeting(https://github.com/apache/apisix/discussions/5811#discussioncomment-1850338), it is best way.

**Reply by juzhiyuan** [reply]
See https://github.com/apache/apisix/discussions/5873

**Comment by tzssangglass** [comment]
@juzhiyuan do we need to talk about this at the next community meeting?

**Reply by juzhiyuan** [reply]
Good idea, I will try to invite @symsimmy 🙈 to join our next meeting :)

**Reply by juzhiyuan** [reply]
See https://docs.qq.com/doc/DTktRakxqaUFFWndz and https://github.com/apache/apisix/discussions/5873

**Reply by symsimmy** [reply]
ok，i will join the meeting to discuss details.

**Reply by juzhiyuan** [reply]
Hi @symsimmy, if possible, you could add my WeChat **JuShaoyao** as well :)
